delivery_product_restriction/models/delivery_carrier.py

```diff
@@ -0,0 +1,29 @@
+# Copyright 2022 Coop IT Easy SCRLfs
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class DeliveryCarrier(models.Model):
+
+    _inherit = "delivery.carrier"
+
+    def _can_be_used_to_deliver_products(self, products):
+        self.ensure_one()
+        return all(
+            not product.restrict_delivery_carrier_to
+            or self in product.restrict_delivery_carrier_to
+            for product in products
+        )
+
+    def available_carriers(self, partner, products=None):
+        delivery_carriers = self
```

Comment:
This is not enough if you want to eliminate side effects:
```py
>>> foo = [1, 2, 3]
>>> bar = foo
>>> bar.pop()
3
>>> foo
[1, 2]
>>> bar
[1, 2]
```

Reply:
I know it's not a deep copy, just a ref copy. There will be no side effect as `filtered` method are not in place method, but only return the filtered recordset.

I need to put all the current delivery in a variable, and then I filtered them one or two times depending on the value of `partner` and `products`.

May be there exist a better way to write that, but I have'nt search a lot.

Reply:
Indeed you're right. 👍